### ISSUE TITLE
prefer enumerations over booleans

### ISSUE THREAD

**Issue by JLindenberg** [SAP/styleguides#14]
maybe ABAP was more modern than other languages in not providing booleans, and with the addition of abap_bool became as worse as others?

There is a debate about using booleans outside, e.g.

http://jlebar.com/2011/12/16/Boolean_parameters_to_API_functions_considered_harmful..html 
https://silkandspinach.net/2004/07/15/avoid-boolean-parameters/
http://www.beyondcode.org/articles/booleanVariables.html (2003)
Reliable Software Technologies - Ada-Europe 2001: 6th Ada-Europe (https://books.google.de/books?id=E0aPZtAMJBQC&pg=PA50&lpg=PA50&dq=boolean+considered+harmful&source=bl&ots=5eXX4ZIMbS&sig=ACfU3U0sAXNMB3Aa-0Y2lo2OF1ua_mekcA&hl=de&sa=X&ved=2ahUKEwjFvay9uP_hAhUB3KQKHahwBrs4FBDoATADegQICRAB#v=onepage&q=boolean%20considered%20harmful&f=false)

In fact it is really hard to date it back as google prefers newer webpages over older…

Can we please add a section “Don´t use booleans where you don´t know it is binary forever?” to https://github.com/SAP/styleguides/blob/master/clean-abap/CleanABAP.md#booleans ?
Not using booleans but enumerations usually improves readability of code significantly unless the domain is in fact binary.
Cheers, Joachim

**Comment by HrFlorianHoffmann** [comment]
The section [_Split method instead of Boolean parameter_](https://github.com/SAP/styleguides/blob/master/clean-abap/CleanABAP.md#split-method-instead-of-boolean-input-parameter) partially covers this. A little unsure whether linking there will do, or whether we should add a new small sub-section on the choice of data type.

**Comment by JLindenberg** [comment]
I´d definitely prefer banning booleans in the first place - as a most likely inadequate data type. If then you end up with an enumeration Do.This or Do.That, then you know how to proceed..

**Comment by HrFlorianHoffmann** [comment]
Added a section [_Use Booleans wisely_](https://github.com/SAP/styleguides/blob/master/clean-abap/CleanABAP.md#use-booleans-wisely) that I think closes this issue.

**Comment by JLindenberg** [comment]
good start, but...
- the link to "Split methods..." is broken (mind the s)
- references 2 & 3 would better fit to "Split methods.." rather than "Use Booleans wisely"
- I am missing the point that readability of booleans is often worse than enumerations even if the domain is binary

**Comment by HrFlorianHoffmann** [comment]
There are several reasons why I didn't comment on the readability so far.

For once, this affects other languages far more than ABAP, because ABAP usually requires you to repeat the parameter names:

```Java
 // Java
updateDocument(true, true, false);
```

while

```ABAP
" ABAP
update_document( synchronous = abap_true
                 with_commit = abap_true
                 write_logs  = abap_false ).
```

ABAP becomes ambiguous only when you're down to one parameter and you can ommit the parameter name:

```ABAP
update( abap_true ).
```

Although this issue frequently occurs with Booleans, it's not caused by them. The issue is in fact caused by the unavailability of the parameter names in the method call, and can happen with all data types. In most cases, enumerations are not even an option:

```Java
// Java
someString.substring(0, 3);  // what's 3? length? end index?
sendEmail("", "", "", "");  // title, body, recipient, sender - or the other way round?
assertEquals(a, b);  // which one is the expected value? a or b?
```

In return, it's not necessary to use a different data type to provide a meaningful enumeration:

```ABAP
" ABAP
CONSTANTS:
  BEGIN OF save_mode,
    with_commit TYPE abap_bool VALUE abap_true,
    without_commit TYPE abap_bool VALUE abap_false,
  END OF save_mode.

update( save_mode-with_commit ).
```

Over the years, most IDEs added the feature to show parameter names inline without them actually being present in the code. This greatly reduces this issue's impact. As far as I know, the ABAP Development Tools do not (yet) support this feature.

In summary, yes, I agree that it's a good idea to challenge whether a Boolean is the right choice, but this ambiguous-call argument is a rather weak one.

**Comment by pokrakam** [comment]
> Over the years, most IDEs added the feature to show parameter names inline without them actually being present in the code. This greatly reduces this issue's impact. As far as I know, the ABAP Development Tools do not (yet) support this feature.

If I understood correctly, you can do this with \<alt\>-F2 from anywhere inside a method.

**Comment by HrFlorianHoffmann** [comment]
That is Eclipse's _Element Info', a quick preview of the method signature in a separate window.

What I mean are "Parameter Hints", as found [for example in IntelliJ](https://www.jetbrains.com/help/rider/Inline_Parameter_Name_Hints.html):

![ScreenshotOfInlinedParameterNamesInIntelliJ](https://user-images.githubusercontent.com/24736032/57308401-979dc900-70e6-11e9-9442-64f52d1ff05b.PNG)

**Comment by pokrakam** [comment]
Ah, nice!
But I guess since parameter names are required if more than one, then it is less of a priority for ABAP.

**Comment by JLindenberg** [comment]
agreed - ABAP is better by naming arguments. 
> ```java
> sendEmail("", "", "", "");  // title, body, recipient, sender - or the other way round?
> ```
actually signatures like these could use more specific types rather then generic types. E.g. recipient and sender could be a class that also provides validation in addition to having a string representation. Is your body html, rtf, or text? Lazy coders.. 

> ```java
> // Java
> someString.substring(0, 3);  // what's 3? length? end index?
> assertEquals(a, b);  // which one is the expected value? a or b?
> ```
no idea.

**Comment by HrFlorianHoffmann** [comment]
After the fixes in https://github.com/SAP/styleguides/pull/25, is there still something we should change?
